Serialization documentation

georgeDaskalakis opened the pull request.

georgeDaskalakis commented on Serialization.md:
```diff
@@ -0,0 +1,184 @@
+
+## Xml Attribute Mapping
+
+The functionality of mapping to specific xml attributes is handled by AttributeMappers that are added to the AttributeMapperRepository either in [OSPSuiteXmlSerializerRepository](https://github.com/Open-Systems-Pharmacology/OSPSuite.Core/blob/develop/src/OSPSuite.Core/Serialization/Xml/OSPSuiteXmlSerializerRepository.cs) in OSPSuite.Core or in the individual solutions in [PKSimXmlSerializerReposi ...
```
I understand that the MoBi file path could change with OSMOSES- but generally the file paths for serilaization of projects would remain pretty constant I guess and I find the link to github helpful. btw @msevestre why is the MoBiXmlSerializerRepository.cs in Presentation?

Yuri05 commented on Serialization.md:
````diff
@@ -0,0 +1,184 @@
+            .
+         </Formulas>
+```
+The above excerpt is a simple PK-Sim project, and more specifically the [S1_concentrBased.pkml](https://github.com/Open-Systems-Pharmacology/PK-Sim/blob/develop/tests/PKSim.Tests/Data/S1_concentrBased.pkml) that is part of the test data of PKSim codebase.
````
"simple PK-Sim project" - would replace this with something like "simple simulation created in PK-Sim and exported to pkml"
(because PK-Sim projects are databases and not xmls/pkmls)

georgeDaskalakis replied:
corrected

Yuri05 commented on Serialization.md:
````diff
@@ -0,0 +1,184 @@
+}
+```
+
+Of if your new class implements an interface that already has an abstract serializer, it is that one that you will need to extend. For example if you are writing a new Building Block class (that would be implementing the interface IBuildingBlock), the serializer signature would be  
````
"Of if" - did you men "Or if"?

georgeDaskalakis replied:
corrected

Yuri05 commented on Serialization.md:
```diff
@@ -0,0 +1,184 @@
+
+# Map(...)
+
+When you want to serialize a property of a class and a serializer already exists for the type of property you want to serialize (as is f.e. for string, int and the other basic types, or in case there is a serializer already written for this object in the solution), you only need to use Map function like in the above example with `Map(x => x.Name);`. You  do not need to explicitly define what need ...
```
here and below: not sure if "f.e." is the common abbreviation, I am used to "e.g."

georgeDaskalakis replied:
I replaced the occurences with "e.g."

Yuri05 commented on Serialization.md:
````diff
@@ -0,0 +1,184 @@
+```
+
+In such a case the `Datarepository` that is `ObservedData` also exists in the Project and gets serialized and deserialized separately. We would not like to keep multiple copies of the same object in our project file, and therefore what we are going to write in the xml is a reference to the `ObservedData`. It is important though in this case to keep in mind than when deserializing we have to make  ...
+
````
Example code for `MapReference `missing?

georgeDaskalakis replied:
WeightedObservedDataXmlSerializer code snipet added

Yuri05 commented on Serialization.md:
```diff
@@ -0,0 +1,184 @@
+# TypedSerialize(...) - TypedDeserialize(...)
+
+In case the above functionalities do not cover your use case, you can also use `TypedSerialize(TObject objectToSerialize, TContext context)` and `TypedDeserialize(TObject objectToDeserialize, XElement outputToDeserialize, TContext context)` to be able to specify the actions that should happen before and after (de)serialization. This is the functionality used f.e. for the (de)serialization of the  ...
+
```
Also here: would be nice to have a code snippet for TypedSerialize / TypedDeserialize

georgeDaskalakis commented on Serialization.md:
```diff
@@ -0,0 +1,184 @@
+
+Now let's talk a bit about the mapping a classes properties in the `PerformMapping()` override. The most frequent use cases would be:
+
+# Map(...)
```
also, I have not included the full signature of the function in the titles because I think that would make the documentation less readable.

msevestre commented:
That looks good to me
I am wondering if we should make it clear that the typical use case over 90% is super dumb end does not require any thinking

Also it is not quite true that the referenced object need to be deserialized before. They need to be deserialized either BEFORE or withing the same serialization /deserialization action (so if A references B and both A and B are withing the root note <ROOT> <A/><B></ROOT> and we deserialize ROOT, this will be also fine

georgeDaskalakis commented:
> That looks good to me I am wondering if we should make it clear that the typical use case over 90% is super dumb end does not require any thinking
> 
> Also it is not quite true that the referenced object need to be deserialized before. They need to be deserialized either BEFORE or withing the same serialization /deserialization action (so if A references B and both A and B are withing the root note and we deserialize ROOT, this will be also fine

OK, I added a small segment to clarify that usually no more than some Map()/MapEnumerable() calls will be necessary. Also corrected the part about the serialization order.

msevestre commented:
awesome

rwmcintosh commented:
Great content!
